File: src/main/java/io/klbz/curie/Conversion.java
package io.klbz.curie;

import static io.klbz.curie.Either.right;
import static io.klbz.curie.Unit.unit;

public final class Conversion {
	private Conversion(){}

	public static <T> Either<T, Unit> maybeToEither(Maybe<T> maybe){
		return maybe
				.map(Either::<T, Unit>left)
				.collapse(right(unit()));
	}

	public static <L, R> Pair<Maybe<L>, Maybe<R>> eitherToPair(Either<L, R> either) {
		return Pair
				.of(either.isolateL())
				.and(either.isolateR());
	}
}

File: src/main/java/io/klbz/curie/Conversion.kt
package io.klbz.curie

import io.klbz.curie.Either.Companion.right
import io.klbz.curie.Unit.Companion.unit
import java.util.function.Function as fn

object Conversion {
    fun <T> maybeToEither(maybe: Maybe<T>) =
            maybe.map<Either<T, Unit>>(fn { Either.left(it) })
                    .collapse(right(unit()))

    fun <L, R> eitherToPair(either: Either<L, R>) =
            Pair.of(either.isolateL())
                    .and(either.isolateR())
}
